alibi_detect/od/pytorch/ensemble.py

```diff
@@ -0,0 +1,341 @@
+
+        Parameters
+        ----------
+        x
+            `torch.Tensor` being transformed.
+
+        Returns
+        -------
+        the transformed `torch.Tensor`.
+        """
+        if not torch.jit.is_scripting():
+            self.check_fitted()
+        return self._transform(x)
+
+
+class PValNormalizer(BaseFittedTransformTorch):
+    def __init__(self):
+        """Maps scores to there p-values.
+
+        Needs to be fit (see :py:obj:`~alibi_detect.od.pytorch.ensemble.BaseFittedTransformTorch`).
+        Returns the proportion of scores in the reference dataset that are greater than the score of
+        interest. Output is between ``1`` and ``0``. Small values are likely to be outliers.
+        """
+        super().__init__()
+        self.val_scores = None
+
+    def _fit(self, val_scores: torch.Tensor) -> PValNormalizer:
+        """Fit transform on scores.
+
+        Parameters
+        ----------
+        val_scores
+            score outputs of ensemble of detectors applied to reference data.
+        """
+        self.val_scores = val_scores
+        return self
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Transform scores to 1 - p-values.
+
+        Parameters
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of 1 - p-values.
+        """
+        p_vals = (
+                1 + (scores[:, None, :] < self.val_scores[None, :, :]).sum(1)
+            )/(len(self.val_scores)+1)
+        return 1 - p_vals
+
+
+class ShiftAndScaleNormalizer(BaseFittedTransformTorch):
+    def __init__(self):
+        """Maps scores to their normalised values.
+
+        Needs to be fit (see :py:obj:`~alibi_detect.od.pytorch.ensemble.BaseFittedTransformTorch`).
+        Subtracts the dataset mean and scales by the standard deviation.
+        """
+        super().__init__()
+        self.val_means = None
+        self.val_scales = None
+
+    def _fit(self, val_scores: torch.Tensor) -> ShiftAndScaleNormalizer:
+        """Computes the mean and standard deviation of the scores and stores them.
+
+        Parameters
+        ----------
+        val_scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+        """
+        self.val_means = val_scores.mean(0)[None, :]
+        self.val_scales = val_scores.std(0)[None, :]
+        return self
+
+    def _transform(self, scores: torch.Tensor) -> torch.Tensor:
+        """Transform scores to normalised values. Subtracts the mean and scales by the standard deviation.
+
+        Parameters
+        ----------
+        scores
+            `Torch.Tensor` of scores from ensemble of detectors.
+
+        Returns
+        -------
+        `Torch.Tensor` of normalised scores.
+        """
+        return (scores - self.val_means)/self.val_scales
```

Comment:
Nitpick: in a few places there's inconsistencies around whitespace around binary operators like these (e.g. compare the minus sign vs division here).

Reply:
This is due to flake8 default error set being: `E121,E123,E126,E226,E24,E704,W503,W504` including `E226` - `missing whitespace around arithmetic operator`. This will be changed in #468 so I'm not sure how we should proceed. Shall we wait until this PR is merged and then update #468? Or fix these issues here but just in the relevant kNN files? It seems more principled to do the fix in #468?